File: src/server/nlp/crisis.ts
import type { EnrichedPost, SentimentLabel } from './types';
import type { AnalyzeResponse } from '../../shared/types';

export const RISK_KEYWORD_LIST = [
  'scam', 'fraud', 'refund', 'lawsuit', 'boycott', 'unsafe', 'danger', 'crisis',
  'outage', 'hack', 'breach', 'leak', 'recall', 'warning', 'complaint', 'delay',
  'broken', 'misleading', 'overpriced', 'ripoff', 'disaster', 'unacceptable',
];

export const extractRiskKeywords = (text: string): string[] => {
  const tokens = new Set(text.split(/\s+/));
  return RISK_KEYWORD_LIST.filter((word) => tokens.has(word));
};

export const riskLevelFromScore = (score: number): string => {
  if (score >= 81) return 'Critical';
  if (score >= 61) return 'High';
  if (score >= 31) return 'Medium';
  return 'Low';
};

export const scoreCrisis = (
  posts: EnrichedPost[],
  spikeDetected: boolean
): AnalyzeResponse['crisis'] => {
  const total = posts.length || 1;
  const negativePosts = posts.filter((p) => p.sentiment === 'negative');
  const negativeRatio = negativePosts.length / total;

  const riskKeywordCount = posts.reduce((sum, p) => sum + p.riskKeywords.length, 0);
  const riskKeywordDensity = riskKeywordCount / total;

  const highImpactNegative = negativePosts.filter(
    (p) => p.score + p.numberOfComments >= 50
  ).length;

  const negativeComponent = negativeRatio * 40;
  const riskComponent = Math.min(riskKeywordDensity * 10, 1) * 25;
  const impactComponent = Math.min(highImpactNegative / 10, 1) * 20;
  const spikeComponent = spikeDetected ? 10 : 0;

  const crisisScore = Math.min(
    100,
    Math.round(negativeComponent + riskComponent + impactComponent + spikeComponent)
  );

  const reasons: string[] = [];
  if (negativeRatio > 0.4) {
    reasons.push(`${Math.round(negativeRatio * 100)}% of posts are negative`);
  }
  if (riskKeywordCount > 0) {
    reasons.push(`${riskKeywordCount} risk keyword mentions detected`);
  }
  if (highImpactNegative > 0) {
    reasons.push(`${highImpactNegative} high-engagement negative posts`);
  }
  if (spikeDetected) {
    reasons.push('Negative sentiment spike detected in recent timeline');
  }
  if (reasons.length === 0) {
    reasons.push('No major crisis signals detected in this sample');
  }

  return {
    crisis_score: crisisScore,
    risk_level: riskLevelFromScore(crisisScore),
    components: {
      negative_ratio: Math.round(negativeComponent),
      risk_keywords: Math.round(riskComponent),
      high_impact_negative: Math.round(impactComponent),
      trend_spike: spikeComponent,
    },
    reasons,
  };
};

export const aggregateRiskKeywords = (
  posts: EnrichedPost[]
): { keyword: string; count: number }[] => {
  const counts = new Map<string, number>();
  for (const post of posts) {
    for (const keyword of post.riskKeywords) {
      counts.set(keyword, (counts.get(keyword) ?? 0) + 1);
    }
  }
  return [...counts.entries()]
    .map(([keyword, count]) => ({ keyword, count }))
    .sort((a, b) => b.count - a.count);
};

export const sentimentCounts = (
  posts: EnrichedPost[]
): Record<SentimentLabel, number> => ({
  positive: posts.filter((p) => p.sentiment === 'positive').length,
  neutral: posts.filter((p) => p.sentiment === 'neutral').length,
  negative: posts.filter((p) => p.sentiment === 'negative').length,
});

File: backend/crisis.py
import math
from typing import TypedDict

from trends import TrendSummary


class CrisisResult(TypedDict):
    crisis_score: int
    risk_level: str
    components: dict[str, float]
    reasons: list[str]


def _impact_score(post: dict) -> float:
    score = post.get("score", 0) or 0
    comments = post.get("num_comments", 0) or 0
    return math.log1p(score) + math.log1p(comments)


def high_impact_posts(posts: list[dict], *, limit: int = 5) -> list[dict]:
    ranked = sorted(posts, key=_impact_score, reverse=True)
    result: list[dict] = []
    for post in ranked[:limit]:
        result.append(
            {
                "id": post.get("id"),
                "title": post.get("title"),
                "author": post.get("author"),
                "subreddit": post.get("subreddit"),
                "score": post.get("score"),
                "num_comments": post.get("num_comments"),
                "sentiment": post.get("sentiment"),
                "risk_keywords": post.get("risk_keywords", []),
                "impact_score": round(_impact_score(post), 2),
                "permalink": post.get("permalink"),
            }
        )
    return result


def calculate_crisis_score(
    posts: list[dict], trends: TrendSummary
) -> CrisisResult:
    total = len(posts) or 1
    neg = sum(1 for p in posts if p.get("sentiment") == "negative")
    neg_pct = neg / total
    risk_total = sum(p.get("risk_keyword_count", 0) for p in posts)
    risk_norm = min(1.0, risk_total / (total * 2))
    neg_high_impact = [
        p for p in posts if p.get("sentiment") == "negative" and _impact_score(p) > 2
    ]
    engagement_ratio = len(neg_high_impact) / total
    trend_spike = 1.0 if trends.get("spike_detected") else 0.0
    if trends.get("avg_negative_ratio", 0) > 0:
        latest_ratio = 0.0
        tl = trends.get("timeline", [])
        if tl:
            latest_ratio = tl[-1].get("negative_ratio", 0)
        trend_spike = max(
            trend_spike,
            min(1.0, latest_ratio / trends["avg_negative_ratio"] - 1)
            if trends["avg_negative_ratio"] > 0
            else 0,
        )

    negative_component = neg_pct * 40
    risk_component = risk_norm * 25
    engagement_component = min(1.0, engagement_ratio * 5) * 20
    trend_component = min(1.0, trend_spike) * 15
    score = int(
        round(
            negative_component
            + risk_component
            + engagement_component
            + trend_component
        )
    )
    score = max(0, min(100, score))

    if score <= 30:
        level = "Low"
    elif score <= 60:
        level = "Medium"
    elif score <= 80:
        level = "High"
    else:
        level = "Critical"

    reasons: list[str] = []
    if neg_pct > 0.5:
        reasons.append(f"Negative sentiment is {neg_pct:.0%} of posts.")
    if risk_total > 0:
        reasons.append(f"Risk keywords appeared {risk_total} times across posts.")
    if neg_high_impact:
        reasons.append(
            f"{len(neg_high_impact)} high-engagement posts carry negative sentiment."
        )
    if trends.get("spike_detected"):
        reasons.append(trends.get("spike_message", "Trend spike detected."))
    if not reasons:
        reasons.append("Discussion appears within normal sentiment bounds.")

    return {
        "crisis_score": score,
        "risk_level": level,
        "components": {
            "negative_sentiment": round(negative_component, 1),
            "risk_keywords": round(risk_component, 1),
            "engagement": round(engagement_component, 1),
            "trend_spike": round(trend_component, 1),
        },
        "reasons": reasons,
    }
